cli/azd/cmd/container.go

```diff
@@ -424,7 +424,11 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 	})
 
 	container.RegisterSingleton(func(ctx context.Context, authManager *auth.Manager) (azcore.TokenCredential, error) {
-		return authManager.CredentialForCurrentUser(ctx, nil)
+		var options = &auth.CredentialForCurrentUserOptions{
+			TenantID: os.Getenv(environment.TenantIdEnvVarName),
```

Comment:
This is a nice hack :)

However, it will not only make the storage sdk client to use the tenant id, but it will also force all other sdk clients to switch.

My proposal:
 - Register a new type and make it exclusive for the remote-env client. Something like `RemoteEnvTokenCredential`
 - Inject this new type for the remote-env flows
 - Add the `tenantId` as part of the remote-env config (from both azd-config and azure.yaml).  Optionally support an ENV-VAR, but I am not 100% sure about it because the other fields are not exposed as ENV VARS...
 - Make a dependency for `RemoteEnvTokenCredential` on the remote-env config and use the tenant when provided.


Additionally, I would consider adding a `azd config set-remote-env` command which would help the user to select or create an storage account and to set the required roles.
This command would persist the tenant and account name, and avoid customer's manual effort to set it up and integrate to azd.

lmk what you think @wbreza

Reply:
I tried to reason about this, and my initial thinking was that since the [authentication SDK uses the same environment variable by default](https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication?tabs=bash#-option-1-define-environment-variables), it would be OK.

I was on mobile and in a bit of a hurry when I read this comment, so I missed parts of what you were saying. After thinking it through for a while, though, I came to realize that there's no basis for assuming the remote env storage resides under the same tenant as the deployed environment--which I think is what you're also saying with this:

> Add the tenantId as part of the remote-env config (from both azd-config and azure.yaml). Optionally support an ENV-VAR, but I am not 100% sure about it because the other fields are not exposed as ENV VARS...

I'd argue that the env variable should also be supported, because it's the default for all (?) the token client libraries, and as such is implicitly supported by a lot of tools. :)

In any case, let me see if I can figure out how to apply your suggestions and learn a bit of golang while I'm at it.

Reply:
I'd also wait a little bit for @wbreza , @rytmis .  He's currently on vacation.  

He designed the remote-env and knows better about what the scope of the feature is. 

As you figured, my comments were around a change which enables folks with multi-tenant accounts to set the tenant where the storage account is.

> because it's the default for all (?) the token client libraries,

client libraries use [azure identity](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/azidentity#azure-identity-client-module-for-go) , which provide a set of different credentials.  Some of those credentials use the `AZURE_TENANT_ID`, but not all of them.
In case of `azd`, it uses the [azure developer cli credential](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/azure_developer_cli_credential.go) which depends on what was used when you run `azd auth login`.
If you run `azd auth login --help`, you would see the supported login options:

```
Log in to Azure.

Usage
  azd auth login [flags]

Flags
        --check-status                          : Checks the log-in status instead of logging in.
        --client-certificate string             : The path to the client certificate for the service principal to authenticate with.
        --client-id string                      : The client id for the service principal to authenticate with.
        --client-secret string                  : The client secret for the service principal to authenticate with. Set to the empty string to read the value from the console.
        --docs                                  : Opens the documentation for azd auth login in your web browser.
        --federated-credential-provider string  : The provider to use to acquire a federated token to authenticate with.
    -h, --help                                  : Gets help for login.
        --redirect-port int                     : Choose the port to be used as part of the redirect URI during interactive login.
        --tenant-id string                      : The tenant id or domain name to authenticate with.
        --use-device-code                       : When true, log in by using a device code instead of a browser.
```

As you can see, there is a `--tenant-id` flag that you can use.
I would expect, on your case, that if you run `azd auth login --tenant-id <your tenant id>` and then try your remote-env config, it would work. But this would alzo make `azd` to only show subscriptions from that tenant-id you selected.

That is also one of the reasons why I don't know if using env-vars would be a good idea. What would it happen if we do `azd auth login --tenant-id fooA` and then set the AZURE_TENANT_ID env var to `fooB`.

Let me tag @ellismg here, as well, as this is also related to `azd login`

Reply:
> I would expect, on your case, that if you run azd auth login --tenant-id <your tenant id> and then try your remote-env config, it would work. But this would alzo make azd to only show subscriptions from that tenant-id you selected.

So did I, but it didn't work. I tried various ways of specifying the tenant, but none of them had any effect--and tracing the code that creates the credential, it looked to me as if it would always pick the user's home tenant, since the options instance was nil. That's why I chose this path as the first stab at a solution. 

I think I've got an idea of how to do what you suggested, so I'll play around with it just to get a hang of things 'til we get the comments from the more informed folks. 😊